File: comment.py
'''
CommentNode
    comment
    parent
    children_list
'''
from __future__ import annotations

class Comment:
    id: str

class CommentNode:
    comment: Comment
    parent: CommentNode
    children_list: list[CommentNode]

    def __init__(self, comment: Comment):
        self.comment = comment
        self.parent = None
        self.children_list = []

def build_tree(comment_list):
    lookup_map = {}
    return_list = []

    for comment in comment_list:
        this_comment = Comment()
        this_comment.id = comment[0]
        new_node = CommentNode(this_comment)
        lookup_map[comment[0]] = new_node

        if not comment[1]:
            return_list.append(new_node)

    for comment in comment_list:
        if comment[1]:
            this_list = lookup_map[comment[1]].children_list or []
            this_list.append(lookup_map[comment[0]])
            lookup_map[comment[1]].children_list = this_list
            lookup_map[comment[0]].parent = lookup_map[comment[1]]
    
    return return_list


def print_tree(comment_node_list):
    for comment_node in comment_node_list:
        print_node(comment_node, 0) 

def print_node(node, level):
    print("\t"*level, node.comment.id)

    for child in node.children_list:
        print_node(child, level + 1)


comment_list = [("1", None),
                ("2", None),
                ("3", None),
                ("4", "1"),
                ("5", "1"),
                ("6", "4"),
                ("7", "5"),
                ("8", "7"),
                ]

print(build_tree(comment_list))
print_tree(build_tree(comment_list))

File: comment.go
package main

// run this with "go run find_chain.go"
import (
	"fmt"
)

type Comment struct {
	id      string
	comment string
}

type CommentNode struct {
	comment       Comment
	parent        *CommentNode
	children_list []*CommentNode
}

type Pair struct {
	a Comment
	b string
}

func main() {

	pair1 := Pair{Comment{id: "1"}, ""}
	pair2 := Pair{Comment{id: "2"}, ""}
	pair3 := Pair{Comment{id: "3"}, ""}
	pair4 := Pair{Comment{id: "4"}, "1"}
	pair5 := Pair{Comment{id: "5"}, "1"}
	pair6 := Pair{Comment{id: "6"}, "4"}
	pair7 := Pair{Comment{id: "7"}, "5"}

	fmt.Println(build_tree([]Pair{pair1, pair2, pair3, pair4, pair5, pair6, pair7}))
	print_tree(build_tree([]Pair{pair1, pair2, pair3, pair4, pair5, pair6, pair7}))
}

func build_tree(arr []Pair) []CommentNode {
	var final_list []CommentNode = []CommentNode{}
	var node_lookup = make(map[string]*CommentNode)

	for idx := 0; idx < len(arr); idx++ {
		var this_node CommentNode = CommentNode{comment: arr[idx].a}
		this_node.comment = arr[idx].a
		node_lookup[arr[idx].a.id] = &this_node
	}

	for idx := 0; idx < len(arr); idx++ {
		var this_node = node_lookup[arr[idx].a.id]

		if arr[idx].b != "" {
			var parent_node = node_lookup[arr[idx].b]

			this_node.parent = parent_node
			parent_node.children_list = append(parent_node.children_list, this_node)
			node_lookup[arr[idx].b] = parent_node
		}
	}

	for idx := 0; idx < len(arr); idx++ {
		if arr[idx].b == "" {
			final_list = append(final_list, *node_lookup[arr[idx].a.id])
		}
	}

	return final_list
}

func print_tree(arr []CommentNode) {
	for idx := 0; idx < len(arr); idx++ {
		print_node(arr[idx], 0)
	}
}

func print_node(node CommentNode, level int) {
	for idx := 0; idx < level; idx++ {
		fmt.Print("\t")
	}

	fmt.Println(node.comment.id)

	for idx := 0; idx < len(node.children_list); idx++ {
		print_node(*node.children_list[idx], level+1)
	}
}
